File: DSAA Interactive Dashboard/p6.c
#include <stdio.h>
#include <stdlib.h>

// ANSI Color Codes for high-tech look
#define RESET   "\033[0m"
#define CYAN    "\033[1;36m"
#define GREEN   "\033[1;32m"
#define RED     "\033[1;31m"
#define AMBER   "\033[1;33m"
#define PURPLE  "\033[1;35m"

struct Node {
    int data;
    struct Node* next;
};

struct Queue {
    struct Node *front, *rear;
};

void printHeader(char* title) {
    printf(PURPLE "\n==================================================\n");
    printf("  [#] %s\n", title);
    printf("==================================================\n" RESET);
}

void printUserGuide() {
    printf(CYAN "\n==================================================\n");
    printf("         CRASH COURSE: HOW TO USE THIS SYSTEM\n");
    printf("==================================================\n" RESET);
    printf(" Think of this program as a lineup of customers connected dynamically via pointers.\n\n");
    printf(AMBER " [STEP 1]" RESET " Interact with the core kernel functions below.\n");
    printf(AMBER " [STEP 2]" RESET " Observe your dynamic FIFO (First-In, First-Out) tracking:\n");
    printf("   • " GREEN "Option 1 (Enqueue):" RESET " Appends a new node to the tail end of the [REAR] chain.\n");
    printf("   • " GREEN "Option 2 (Dequeue):" RESET " Detaches and frees the oldest surviving element at [FRONT].\n");
    printf("   • " GREEN "Option 3 (View):" RESET "    Outputs sequential address links maps terminated with a NULL string.\n");
    printf(PURPLE "==================================================\n\n" RESET);
    printf("Press " AMBER "[ENTER]" RESET " to initialize dynamic queue interface...");
    while(getchar() != '\n'); // Clear buffer
    getchar();
}

void init(struct Queue* q) { 
    q->front = q->rear = NULL; 
}

int isEmpty(struct Queue* q) { 
    return q->front == NULL; 
}

void enqueue(struct Queue* q) {
    int val;
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (!newNode) { 
        printf(RED "\n [!] FATAL: HEAP MEMORY EXHAUSTED\n" RESET); 
        return; 
    }
    
    printf(AMBER " [?] Input Value to Enqueue: " RESET);
    scanf("%d", &val);
    
    newNode->data = val;
    newNode->next = NULL;
    
    if (q->rear == NULL) {
        q->front = q->rear = newNode;
    } else {
        q->rear->next = newNode;
        q->rear = newNode;
    }
    printf(GREEN "  (+) Success: Dynamic node allocated and linked at rear pointer.\n" RESET);
}

void dequeue(struct Queue* q) {
    if (isEmpty(q)) { 
        printf(RED "\n [!] UNDERFLOW: LINKED QUEUE CONTAINS ZERO ELEMENTS\n" RESET); 
        return; 
    }
    
    struct Node* temp = q->front;
    int data = temp->data;
    
    q->front = q->front->next;
    
    if (q->front == NULL) {
        q->rear = NULL;
    }
    
    free(temp);
    printf(RED "  (-) Dropped: Front cell popped -> data [%d] unlinked & freed\n" RESET, data);
}

void display(struct Queue* q) {
    printHeader("DYNAMIC QUEUE MEMORY VISUALIZATION");
    if (isEmpty(q)) { 
        printf(RED "  [ !! SYSTEM EMPTY REGISTER FLUSH !! ] \n" RESET); 
        return; 
    }
    
    struct Node* temp = q->front;
    printf("  ");
    while (temp) {
        if (temp == q->front) 
            printf(AMBER "[%d]" RESET " (FRONT) -> ", temp->data);
        else if (temp == q->rear)
            printf(PURPLE "[%d]" RESET " (REAR) -> ", temp->data);
        else
            printf(CYAN "[%d]" RESET " -> ", temp->data);
            
        temp = temp->next;
    }
    printf(PURPLE "NULL\n" RESET);
}

int main() {
    struct Queue q;
    init(&q);
    int choice;

    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif

    printUserGuide();

    printf(CYAN "==================================================\n");
    printf("     [M] DYNAMIC LINKED QUEUE INTERFACE [M]\n");
    printf("==================================================\n" RESET);

    while (1) {
        printf(CYAN "\n [MAIN MENU]\n" RESET);
        printf("  1.  (+) Enqueue Allocation (Insert Element)\n");
        printf("  2.  (-) Dequeue Destruction (Remove Element)\n");
        printf("  3.  (#) Traverse Node Layout Sequence\n");
        printf("  4.  [$] Run Pointer Diagnostic Status\n");
        printf("  5.  (X) Terminate Control Kernel Execution\n");
        printf(AMBER " >>> Select Target Action Option: " RESET);
        
        if (scanf("%d", &choice) != 1) break;

        switch (choice) {
            case 1: enqueue(&q); break;
            case 2: dequeue(&q); break;
            case 3: display(&q); break;
            case 4:
                printHeader("HEAP STRUCT POINTER DIAGNOSTIC REPORT");
                printf("  [-] Front Structural Pointer:   %p\n", (void*)q.front);
                printf("  [-] Rear Structural Pointer :   %p\n", (void*)q.rear);
                printf("  [-] Dynamic Memory State Empty: %s\n", isEmpty(&q) ? RED "TRUE (NULL)" RESET : GREEN "FALSE (Active Nodes)" RESET);
                break;
            case 5: 
                printf(RED "\n [!] CORE KERNEL POWER DOWN INTERRUPT ISSUED...\n" RESET);
                exit(0);
            default: 
                printf(RED " [!] WARNING: Operational instruction unrecognized.\n" RESET);
        }
    }
    return 0;
}

File: DSAA Interactive Dashboard/p6.py
import os
import sys

RESET = "\033[0m"
CYAN = "\033[1;36m"
GREEN = "\033[1;32m"
RED = "\033[1;31m"
AMBER = "\033[1;33m"
PURPLE = "\033[1;35m"

class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

def print_header(title):
    print(f"{PURPLE}\n==================================================")
    print(f"  [#] {title}")
    print(f"=================================================={RESET}")

def print_user_guide():
    print(f"{CYAN}\n==================================================")
    print("         CRASH COURSE: HOW TO USE THIS SYSTEM")
    print(f"=================================================={RESET}")
    print(" Think of this program as a lineup of customers connected dynamically via pointers.\n")
    print(f"{AMBER} [STEP 1]{RESET} Interact with the core kernel functions below.")
    print(f"{AMBER} [STEP 2]{RESET} Observe your dynamic FIFO (First-In, First-Out) tracking:")
    print(f"   • {GREEN}Option 1 (Enqueue):{RESET} Appends a new node to the tail end of the [REAR] chain.")
    print(f"   • {GREEN}Option 2 (Dequeue):{RESET} Detaches and frees the oldest surviving element at [FRONT].")
    print(f"   • {GREEN}Option 3 (View):{RESET}    Outputs sequential address links maps terminated with a NULL string.")
    print(f"{PURPLE}==================================================\n{RESET}")
    input(f"Press {AMBER}[ENTER]{RESET} to initialize dynamic queue interface...")

def is_empty(q):
    return q.front is None

def enqueue(q):
    try:
        val = int(input(f"{AMBER} [?] Input Value to Enqueue: {RESET}"))
    except ValueError:
        print(f"{RED} [!] ERROR: Invalid processing frame data frame.{RESET}")
        return

    new_node = Node(val)
    if q.rear is None:
        q.front = q.rear = new_node
    else:
        q.rear.next = new_node
        q.rear = new_node
    print(f"{GREEN}  (+) Success: Dynamic node allocated and linked at rear pointer.{RESET}")

def dequeue(q):
    if is_empty(q):
        print(f"{RED}\n [!] UNDERFLOW: LINKED QUEUE CONTAINS ZERO ELEMENTS{RESET}")
        return
    
    temp = q.front
    data = temp.data
    q.front = q.front.next
    
    if q.front is None:
        q.rear = None
        
    print(f"{RED}  (-) Dropped: Front cell popped -> data [{data}] unlinked & freed{RESET}")

def display(q):
    print_header("DYNAMIC QUEUE MEMORY VISUALIZATION")
    if is_empty(q):
        print(f"{RED}  [ !! SYSTEM EMPTY REGISTER FLUSH !! ] \n{RESET}")
        return
        
    temp = q.front
    print("  ", end="")
    while temp:
        if temp == q.front:
            print(f"{AMBER}[{temp.data}]{RESET} (FRONT) -> ", end="")
        elif temp == q.rear:
            print(f"{PURPLE}[{temp.data}]{RESET} (REAR) -> ", end="")
        else:
            print(f"{CYAN}[{temp.data}]{RESET} -> ", end="")
        temp = temp.next
    print(f"{PURPLE}NULL{RESET}")

def main():
    os.system('cls' if os.name == 'nt' else 'clear')
    print_user_guide()

    q = Queue()

    print(f"{CYAN}==================================================")
    print("     [M] DYNAMIC LINKED QUEUE INTERFACE [M]")
    print(f"=================================================={RESET}")

    while True:
        print(f"{CYAN}\n [MAIN MENU]{RESET}")
        print("  1.  (+) Enqueue Allocation (Insert Element)")
        print("  2.  (-) Dequeue Destruction (Remove Element)")
        print("  3.  (#) Traverse Node Layout Sequence")
        print("  4.  [$] Run Pointer Diagnostic Status")
        print("  5.  (X) Terminate Control Kernel Execution")
        
        try:
            choice = int(input(f"{AMBER} >>> Select Target Action Option: {RESET}"))
        except ValueError:
            print(f"{RED} [!] WARNING: Operational instruction unrecognized.{RESET}")
            continue

        if choice == 1: enqueue(q)
        elif choice == 2: dequeue(q)
        elif choice == 3: display(q)
        elif choice == 4:
            print_header("HEAP STRUCT POINTER DIAGNOSTIC REPORT")
            # Python extracts identity hashes as hex equivalents for raw pointer simulation
            f_ptr = f"0x{id(q.front):02X}" if q.front else "NIL"
            r_ptr = f"0x{id(q.rear):02X}" if q.rear else "NIL"
            print(f"  [-] Front Structural Pointer:   {f_ptr}")
            print(f"  [-] Rear Structural Pointer :   {r_ptr}")
            print(f"  [-] Dynamic Memory State Empty: {RED + 'TRUE (NULL)' + RESET if is_empty(q) else GREEN + 'FALSE (Active Nodes)' + RESET}")
        elif choice == 5:
            print(f"{RED}\n [!] CORE KERNEL POWER DOWN INTERRUPT ISSUED...\n{RESET}")
            sys.exit(0)
        else:
            print(f"{RED} [!] WARNING: Operational instruction unrecognized.{RESET}")

if __name__ == "__main__":
    main()